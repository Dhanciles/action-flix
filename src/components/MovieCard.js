import React from 'react'

export const MovieCard = ({title, rating, releaseDate, description}) => {
  return (
    <article>
      <p>{rating}</p>
      <h3>{title}</h3>
      <p>{releaseDate}</p>
      <p>{description}</p>
    </article>
  )
}