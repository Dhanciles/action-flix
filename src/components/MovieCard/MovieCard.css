.movie-card {
  border: 5px solid #8E959C;
  display: flex;
  height: 300px;
  margin-bottom: 25px;
  padding-right: 10px;
  width: 500px;
}

.movie-img {
  height: 300px;
  margin-right: 20px;
  width: 185px;
}

.movie-info-text {
  width: 100%;
}

.movie-icon-container {
  display: flex;
  justify-content: space-between;
  padding: 0 0 5px 0;
}

.movie-rating-chart {
  height: 50px;
  padding-top: 12px;
  width: 50px;
}

.outline-circle,
.circle {
  stroke-width: 3;
}

.outline-circle {
  fill: #8E959C;
  stroke: #fff;
}

.circle {
  fill: none;
  animation: progress 4s ease-out forwards;
  stroke: #76E1EF;
  stroke-linecap: round;
}

@keyframes progress {
  0% {
    stroke-dasharray: 0 100;
  }

  60% {
    stroke-dasharray: 100 100;
  }
}

.movie-rating {
  font-size: 0.8rem;
}

.popcorn-icon {
  height: 60px;
  /*margin-bottom: 10px;*/
}

.movie-title, 
.movie-release {
  padding-left: 5px;
}

.movie-title {
  color: #76E1EF;
  font-size: 1.5rem;
  margin: 0;
}

.movie-release,
.movie-description {
  color: #8E959C;
}

.movie-description {
  border-radius: 10px;
  cursor: default;
  overflow: auto;
  width: 200px;
  height: 100px;
  transition: box-shadow 500ms;
  padding: 5px;
}

.movie-description:hover {
  box-shadow: inset 0 -45px 130px -55px rgba(0, 0, 0, 1);
}

