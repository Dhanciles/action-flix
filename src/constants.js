const movieDataBaseKey =  '8d54c10134c07bfb8e01ded9e30524be';

export {
  movieDataBaseKey
}