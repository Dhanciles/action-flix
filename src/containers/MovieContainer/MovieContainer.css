.movie-container {
  background-color: #323740;
  display: flex;
  justify-content: space-around;
  flex-wrap: wrap;
  margin-top: 25px;
}