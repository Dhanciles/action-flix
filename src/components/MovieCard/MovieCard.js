import React from 'react'
import './MovieCard.css'
import { connect } from 'react-redux'


export const MovieCard = ({title, posterPath, releaseDate, voteAverage, overview, props}) => {
  return (
    <article className='movie-card'>
      <img className='movie-img' src={posterPath} alt='movie poster' />
      <div className='movie-info-text'>
        <div className='movie-icon-container'>
          <p className='movie-rating'>{voteAverage}</p>
          <img src='../images/full-popcorn.svg' alt='full popcorn' onClick={() => props.addFavorite(movieId, props.userId, title, posterPath, releaseDate, voteAverage, overview)} />
        </div>
        <h3 className='movie-title'>{title}</h3>
        <p className='movie-release'>{releaseDate}</p>
        <p className='movie-description'>{overview}</p>
      </div>
    </article>
  )

  export const mapDispatchToProps = (dispatch) => ({
    addFavorite: () => dispatch(postFavorite(movieId, userId, title, posterPath, releaseDate, voteAverage, overview))
  })

  export const mapStateToProps = (state) => ({
    userId: state.userId
  })

  export default connect(mapStateToProps, mapDispatchToProps)(MovieCard)
}