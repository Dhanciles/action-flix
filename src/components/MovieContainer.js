import React from 'react'
import { MovieCard } from './MovieCard'

export const MovieContainer = ({ movies }) => {
  const actionMovies = movies.map(movie => (<MovieCard {...movie} />))

  return (
    <div>
      {actionMovies}
    </div>
  )
}