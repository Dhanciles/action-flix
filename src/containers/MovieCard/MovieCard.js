import React from 'react'
import './MovieCard.css'
import { Component } from 'react'
import { withRouter } from 'react-router-dom'
import { connect } from 'react-redux'
import { postFavorite } from '../../thunks/postFavorite.js'
import { deleteFavorite } from '../../thunks/deleteFavorite.js'
import { toggleFavorite } from '../../actions/index.js'
import PropTypes from 'prop-types'



export class MovieCard extends Component {

  checkSignedIn = () => {
    if (this.props.userId === '') {
      this.props.history.push(`/login`)
    } else {
      const isFavorite = this.props.movies.find(movie => {
        return movie.movieId === this.props.movieId
      })
      if(!isFavorite.favorite) {
        this.props.addFavorite(this.props.movieId, this.props.userId, this.props.title, this.props.posterPath, this.props.releaseDate, this.props.voteAverage, this.props.overview)
        this.props.toggleFavorite(this.props.title)
      } else {
        this.props.toggleFavorite(this.props.title)
        this.props.deleteFavorite(this.props.userId, this.props.movieId)
      }
    }
  }

  render() {
   const favoriteIcon = this.props.favorite ? '../images/full-popcorn.svg' : '../images/empty-popcorn.svg'
    return (
      <article className='movie-card'>
        <img className='movie-img' src={this.props.posterPath} alt='movie poster' />
        <div className='movie-info-text'>
          <div className='movie-icon-container'>
            <svg className='movie-rating-chart' viewBox="0 0 36 36">
              <path className='outline-circle'
              d="M18 2.0845
              a 15.9155 15.9155 0 0 1 0 31.831
              a 15.9155 15.9155 0 0 1 0 -31.831"
              strokeDasharray="100, 100"               
              />
              <path className='circle'
              d="M18 2.0845
              a 15.9155 15.9155 0 0 1 0 31.831
              a 15.9155 15.9155 0 0 1 0 -31.831"
              strokeDasharray={`${this.props.voteAverage * 10}, 100`}               
              />
              <text textAnchor="middle" x="17" y="22" className='movie-rating'>{this.props.voteAverage}</text>
            </svg>
            <img className='popcorn-icon' src={favoriteIcon} alt='full popcorn' onClick={() => this.checkSignedIn()} />
          </div>
          <h3 className='movie-title'>{this.props.title}</h3>
          <p className='movie-release'>{this.props.releaseDate}</p>
          <div className='movie-description'>
            {this.props.overview}
          </div>
        </div>
      </article>
    )
  }
}

MovieCard.propTypes = {
  addFavorite: PropTypes.func.isRequired, 
  toggleFavorite: PropTypes.func.isRequired, 
  deleteFavorite: PropTypes.func.isRequired,
  movies: PropTypes.array.isRequired, 
  userId: PropTypes.number
}


export const mapDispatchToProps = (dispatch) => ({
  addFavorite: (movieId, userId, title, posterPath, releaseDate, voteAverage, overview) => dispatch(postFavorite(movieId, userId, title, posterPath, releaseDate, voteAverage, overview)),
  toggleFavorite: (movieTitle) => dispatch(toggleFavorite(movieTitle)),
  deleteFavorite: (userId, movieId) => dispatch(deleteFavorite(userId, movieId))
})

export const mapStateToProps = (state, props) => ({
  movies: state.movies,
  userId: state.userId,
  movieId: props.movieId, 
  title: props.title, 
  posterPath: props.posterPath, 
  releaseDate: props.releaseDate, 
  voteAverage: props.voteAverage, 
  overview: props.overview, 
  favorite: props.favorite
})

export default connect(mapStateToProps, mapDispatchToProps)(withRouter(MovieCard))

