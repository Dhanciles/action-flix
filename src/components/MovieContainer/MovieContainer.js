import React, { Component } from 'react'
import { Route } from 'react-router-dom'
import MovieCard from '../MovieCard/MovieCard'
import './MovieContainer.css'
import { connect } from 'react-redux'
import { fetchMovies } from '../../thunks/fetchMovies.js'
import { movieDataBaseKey }  from '../../../src/constants.js'
import { getFavorites } from '../../thunks/getFavorites.js'


export class MovieContainer extends Component {
async componentDidMount() {
  let apiKey = movieDataBaseKey
  const url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&with_genres=28`
   await this.props.fetchMovies(url, this.props.userId)
}

render() {
  if (this.props.movies && this.props.location === '/') {
    const actionMovies = this.props.movies.map(movie => (<MovieCard {...movie} />))
    return (
      <div className='movie-container'>
        {actionMovies}
      </div>
    )
  } 
  // else if (this.props.movies && this.props.location === '/favorites') {
  //     const favoriteMovies = favoriteMoviesArr.map((movie) => (<MovieCard {...movie.props} />))

  //     return (
  //       <div className='movie-container'>
  //         {favoriteMovies}
  //       </div>
  //     )
  //   } 
    else {
      return <div>{'error'}</div>
    }
  }
}

export const mapStateToProps = (state) => {
  return {
  userId: state.userId,
  movies: state.movies,
  isLoading: state.isLoading,
  hasErrored: state.hasErrored
}}

export const mapDispatchToProps = (dispatch) => ({
  fetchMovies: (url, userId) => dispatch(fetchMovies(url, userId)),
  getFavorites: (userId) => dispatch(getFavorites(userId)) 
})

export default connect(mapStateToProps, mapDispatchToProps)(MovieContainer)
