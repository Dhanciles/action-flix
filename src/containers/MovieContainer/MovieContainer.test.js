import React from 'react'
import { shallow, mount } from 'enzyme'
import { MovieContainer, mapStateToProps, mapDispatchToProps } from './MovieContainer';
import { fetchMovies } from '../../thunks/fetchMovies.js';
import { MovieCard } from '../MovieCard/MovieCard'
import { movieDataBaseKey }  from '../../../src/constants.js'

describe('MovieContainer', () => {
  let apiKey
  let url
  let mockFetch
  let mockDispatch
  let mockMovies
  let wrapper
  let userId

  beforeEach(() => {
    apiKey = movieDataBaseKey
    url = `https://api.themoviedb.org/3/discover/movie?api_key=${apiKey}&with_genres=28`
    mockFetch = jest.fn()
    mockDispatch = jest.fn()
    mockMovies = [{name: 'Aquaman'}]
    userId = 1
    wrapper = shallow(<MovieContainer getFavorites={jest.fn()} userId={userId} movies={ mockMovies } isLoading={ false } hasErrored={ false } fetchMovies={ mockFetch } />)
  })

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should render a no favorites message if there are no favorite movies on favorites page in props', () => {
    const mockMovies = []
    const wrapper = shallow(<MovieContainer getFavorites={jest.fn()} userId={ userId } movies={ mockMovies } isLoading={ false } hasErrored={ true } fetchMovies={ mockFetch } location={ '/favorites' } />)

    expect(wrapper.find('section').length).toEqual(1)
  })

  it('should render a 404 error if there are no movies in props', () => {
    const mockMovies = []
    const wrapper = shallow(<MovieContainer getFavorites={jest.fn()} userId={ userId } movies={ mockMovies } isLoading={ false } hasErrored={ true } fetchMovies={ mockFetch } />)

    expect(wrapper.find('h3').length).toEqual(1)
  })

  it('should render a loading message if there are no movies in props', () => {
    const mockMovies = []
    const wrapper = shallow(<MovieContainer getFavorites={jest.fn()} userId={ userId } movies={ mockMovies } isLoading={ true } hasErrored={ false } fetchMovies={ mockFetch } />)

    expect(wrapper.find('h1').length).toEqual(1)
  })

  describe('componentDidMount', () => {
    it('should call fetchMovies with the correct params', () => {
      wrapper.instance().componentDidMount()

      expect(mockFetch).toHaveBeenCalledWith(url, userId)      
    })
  })

  it('should render actionMovies when webpage is at home location', () => {
    wrapper = shallow(<MovieContainer location='/' getFavorites={jest.fn()} userId={userId} movies={ mockMovies } isLoading={ false } hasErrored={ false } fetchMovies={ mockFetch }/>)
    expect(wrapper.find('.action-movies').length).toBe(1)
  })

  it('should render favoriteMovies when webpage is at favorites location', () => {
    const mockMovies = [{name: 'Aquaman', favorite: true}, {name: 'Captain America', favorite: true}]
    wrapper = shallow(<MovieContainer location='/favorites' getFavorites={jest.fn()} userId={userId} movies={ mockMovies } isLoading={ false } hasErrored={ false } fetchMovies={ mockFetch }/>)
    
    expect(wrapper.find('.favorite-movies').length).toBe(1)
  })  

  describe('mapStateToProps', () => {
   it('should return an object with the keys of movies, isLoading, and hasErrored', () => {
     const mockState = {
       movies: [{name: 'Aquaman'}],
       isLoading: false,
       hasErrored: false
     }

     const expected = {
       movies: [{name: 'Aquaman'}],
       isLoading: false,
       hasErrored: false
     }

     const result = mapStateToProps(mockState)

     expect(result).toEqual(expected)
   })
 })

  describe('mapDispatchToProps', () => {
   it('should call dispatch with the correct params', () => {
     const mappedProps = mapDispatchToProps(mockDispatch)
     const mockUrl = 'www.getthemovies.com'

     mappedProps.fetchMovies(mockUrl)

     expect(mockDispatch).toHaveBeenCalled
   })

   it('should call dispatch with the correct params', () => {
     const mappedProps = mapDispatchToProps(mockDispatch)
     const mockId = 1

     mappedProps.getFavorites(mockId)

     expect(mockDispatch).toHaveBeenCalled
   })
 })
})