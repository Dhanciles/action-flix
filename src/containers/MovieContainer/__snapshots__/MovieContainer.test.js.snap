// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieContainer should match the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieContainer
    fetchMovies={
      [MockFunction] {
        "calls": Array [
          Array [
            "https://api.themoviedb.org/3/discover/movie?api_key=8d54c10134c07bfb8e01ded9e30524be&with_genres=28",
            1,
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    getFavorites={[MockFunction]}
    hasErrored={false}
    isLoading={false}
    movies={
      Array [
        Object {
          "name": "Aquaman",
        },
      ]
    }
    userId={1}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "error",
    },
    "ref": null,
    "rendered": "error",
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "error",
      },
      "ref": null,
      "rendered": "error",
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
